File: pact-jvm-matchers/src/main/scala/au/com/dius/pact/matchers/PlainTextBodyMatcher.scala
package au.com.dius.pact.matchers

import java.util.Collections

import au.com.dius.pact.model._
import au.com.dius.pact.model.matchingrules.{MatchingRules, RegexMatcher}
import com.typesafe.scalalogging.StrictLogging
import scala.collection.JavaConverters._

class PlainTextBodyMatcher extends BodyMatcher with StrictLogging {

  override def matchBody(expected: HttpPart, actual: HttpPart, allowUnexpectedKeys: Boolean): java.util.List[BodyMismatch] = {
    (expected.getBody.getState, actual.getBody.getState) match {
      case (OptionalBody.State.MISSING, _) => Collections.emptyList()
      case (OptionalBody.State.NULL, OptionalBody.State.PRESENT) => Collections.singletonList(new BodyMismatch(null, actual.getBody.getValue,
        s"Expected empty body but received '${actual.getBody.getValue}'"))
      case (OptionalBody.State.NULL, _) => Collections.emptyList()
      case (_, OptionalBody.State.MISSING) => Collections.singletonList(new BodyMismatch(expected.getBody.getValue, null,
        s"Expected body '${expected.getBody.getValue}' but was missing"))
      case (OptionalBody.State.EMPTY, OptionalBody.State.EMPTY) => Collections.emptyList()
      case (_, _) => compareText(expected.getBody.orElse(""), actual.getBody.orElse(""), expected.getMatchingRules).asJava
    }
  }

  def compareText(expected: String, actual: String, matchers: MatchingRules): List[BodyMismatch] = {
    val regex = matchers.rulesForCategory("body").getMatchingRules.get("$")

    if(regex == null || regex.getRules.isEmpty || !regex.getRules.get(0).isInstanceOf[RegexMatcher]) {
      logger.debug("No regex for " + expected + ", using equality")

      if(expected == actual) {
        return List()
      }

      return List(new BodyMismatch(expected, actual,
        s"Expected body '${expected}' to match '${actual}' using equality but did not match"))
    }

    if (actual.matches(regex.getRules.get(0).asInstanceOf[RegexMatcher].getRegex)) {
      return List()
    }

    List(new BodyMismatch(expected, actual,
      s"Expected body '${expected}' to match '${actual}' using regex '${regex.toString()}' but did not match"))
  }
}

File: core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/PlainTextBodyMatcher.kt
package au.com.dius.pact.core.matchers

import au.com.dius.pact.core.model.OptionalBody
import au.com.dius.pact.core.model.matchingrules.MatchingRules
import au.com.dius.pact.core.model.matchingrules.RegexMatcher
import mu.KLogging

class PlainTextBodyMatcher : BodyMatcher {

  override fun matchBody(
    expected: OptionalBody,
    actual: OptionalBody,
    context: MatchingContext
  ): BodyMatchResult {
    return when {
      expected.isMissing() -> BodyMatchResult(null, emptyList())
      expected.isNull() && actual.isPresent() -> BodyMatchResult(null, listOf(
        BodyItemMatchResult("$",
          listOf(BodyMismatch(null, actual!!.value, "Expected empty body but received '${actual.value}'")))))
      expected.isNull() -> BodyMatchResult(null, emptyList())
      actual.isMissing() -> BodyMatchResult(null, listOf(BodyItemMatchResult("$",
        listOf(BodyMismatch(expected!!.value, null,
          "Expected body '${expected.value}' but was missing")))))
      expected.isEmpty() && actual.isEmpty() -> BodyMatchResult(null, emptyList())
      else -> BodyMatchResult(null,
        compareText(expected.valueAsString(), actual.valueAsString(), context))
    }
  }

  fun compareText(expected: String, actual: String, context: MatchingContext): List<BodyItemMatchResult> {
    val regexMatcher = context.matchers.matchingRules["$"]
    val regex = regexMatcher?.rules?.first()

    if (regexMatcher == null || regexMatcher.rules.isEmpty() || regex !is RegexMatcher) {
      logger.debug { "No regex for '$expected', using equality" }
      return if (expected == actual) {
        listOf(BodyItemMatchResult("$", emptyList()))
      } else {
        listOf(BodyItemMatchResult("$", listOf(BodyMismatch(expected, actual,
          "Expected body '$expected' to match '$actual' using equality but did not match"))))
      }
    }

    return if (actual.matches(Regex(regex.regex))) {
      emptyList()
    } else {
      listOf(BodyItemMatchResult("$", listOf(BodyMismatch(expected, actual,
        "Expected body '$expected' to match '$actual' using regex '${regex.regex}' but did not match"))))
    }
  }

  companion object : KLogging()
}
